File: brute_force.c
#include <stdio.h>
#include <stdbool.h>
#include <time.h>
#include <math.h>

#define REPETICOES 30

bool subset_sum(int set[], int target, int n) {
    if (target == 0)
        return true;
    if (n == 0)
        return false;
    if (set[n - 1] > target)
        return subset_sum(set, target, n - 1);
    return subset_sum(set, target, n - 1) || subset_sum(set, target - set[n - 1], n - 1);
}

int calcular_soma(int set[], int n) {
    int soma = 0;
    for (int i = 0; i < n; i++) {
        soma += set[i];
    }
    return soma;
}

void medir_execucao(double* media, double* desvio, int set[], int n, int target) {
    double tempos[REPETICOES];
    for (int i = 0; i < REPETICOES; i++) {
        clock_t inicio = clock();
        subset_sum(set, target, n);
        clock_t fim = clock();
        tempos[i] = (double)(fim - inicio) / CLOCKS_PER_SEC;
    }

    double soma = 0.0;
    for (int i = 0; i < REPETICOES; i++) soma += tempos[i];
    *media = soma / REPETICOES;

    double variancia = 0.0;
    for (int i = 0; i < REPETICOES; i++)
        variancia += pow(tempos[i] - *media, 2);
    *desvio = sqrt(variancia / REPETICOES);
}

int main() {
    double mean_p, std_p, mean_m, std_m, mean_g, std_g;

    int pequena[] = {1, 2, 3, 4, 5};
    int n_pequena = sizeof(pequena) / sizeof(pequena[0]);
    int alvo_pequena = calcular_soma(pequena, n_pequena) / 2;
    medir_execucao(&mean_p, &std_p, pequena, n_pequena, alvo_pequena);

    int media[15];
    for (int i = 0; i < 15; i++) media[i] = i + 1;
    int alvo_media = calcular_soma(media, 15) / 2;
    medir_execucao(&mean_m, &std_m, media, 15, alvo_media);

    int grande[33];
    for (int i = 0; i < 33; i++) grande[i] = i + 1;
    int alvo_grande = calcular_soma(grande, 33) / 2;
    medir_execucao(&mean_g, &std_g, grande, 33, alvo_grande);

    FILE* arquivo = fopen("results_c.txt", "w");
    if (arquivo != NULL) {
        fprintf(arquivo, "mean-p=%.6f\n", mean_p);
        fprintf(arquivo, "mean-m=%.6f\n", mean_m);
        fprintf(arquivo, "mean-g=%.6f\n", mean_g);
        fprintf(arquivo, "std-p=%.6f\n", std_p);
        fprintf(arquivo, "std-m=%.6f\n", std_m);
        fprintf(arquivo, "std-g=%.6f\n", std_g);
        fclose(arquivo);
        printf("Resultados salvos em data/results_c.txt\n");
    } else {
        printf("Erro ao abrir o arquivo 'data/results_c.txt' para escrita.\n");
    }

    return 0;
}

File: brute_force.py
import os
import time
import numpy as np
import pandas as pd

def subset_sum(set_, target, n=None):
    if n is None:
        n = len(set_)
    if target == 0:
        return True
    if n == 0:
        return False
    if set_[n - 1] > target:
        return subset_sum(set_, target, n - 1)
    return subset_sum(set_, target, n - 1) or subset_sum(set_, target - set_[n - 1], n - 1)

def medir_tempo_execucao(set_, target, rep=30):
    tempos = []
    for _ in range(rep):
        inicio = time.perf_counter()
        subset_sum(set_, target)
        fim = time.perf_counter()
        tempos.append(fim - inicio)
    return {
        "tempo_medio_s": np.mean(tempos),
        "desvio_padrao_s": np.std(tempos)
    }

entrada_pequena = list(range(1, 19))   

alvo_pequeno = sum(entrada_pequena) // 2

entrada_media = list(range(1, 26))        
alvo_medio = sum(entrada_media) // 2

entrada_grande = list(range(1, 34))        
alvo_grande = sum(entrada_grande) // 2

resultado_pequeno = medir_tempo_execucao(entrada_pequena, alvo_pequeno)
resultado_medio = medir_tempo_execucao(entrada_media, alvo_medio)
resultado_grande = medir_tempo_execucao(entrada_grande, alvo_grande)

df_tempos_ajustado = pd.DataFrame([
    {"categoria": "Pequena (n=18)", **resultado_pequeno},
    {"categoria": "Media (n=25)", **resultado_medio},
    {"categoria": "Grande (n=33)", **resultado_grande}
])

df_tempos_ajustado["categoria"] = [
    "Pequena (n=18)",
    "Média (n=25)",
    "Grande (n=33)"
]

os.makedirs("data", exist_ok=True)

caminho_arquivo = "data/results_py.txt"

mean_p = df_tempos_ajustado.loc[0, "tempo_medio_s"]
std_p = df_tempos_ajustado.loc[0, "desvio_padrao_s"]
mean_m = df_tempos_ajustado.loc[1, "tempo_medio_s"]
std_m = df_tempos_ajustado.loc[1, "desvio_padrao_s"]
mean_g = df_tempos_ajustado.loc[2, "tempo_medio_s"]
std_g = df_tempos_ajustado.loc[2, "desvio_padrao_s"]

conteudo_formatado = (
    f"mean-p={mean_p:.6f}\n"
    f"mean-m={mean_m:.6f}\n"
    f"mean-g={mean_g:.6f}\n"
    f"std-p={std_p:.6f}\n"
    f"std-m={std_m:.6f}\n"
    f"std-g={std_g:.6f}\n"
)

with open("data/results_py.txt", "w") as f:
    f.write(conteudo_formatado)


